Bpm improvements

ferranlala opened the pull request:
This fixes [bug #1344727: Allow to change the number of taps for BPM averaging Edit](https://bugs.launchpad.net/mixxx/+bug/1344727).
Now the bpm is calculated via a 80 samples moving interquartile mean which provides an estimation that is robust against outliers, e.g. a missed tap.

ywwg committed:
An anti-aliased whitelabel drawing

esbrandt committed:
Merge split sentences in tooltips

esbrandt committed:
Update translation template. Found 1905 source text(s) (25 new and 1880 already existing). Kept 6 obsolete entries.

Commit:
Change tooltip for filter effect

Owen suggested a better description in the manual. I used it now here as well.

ferranlala committed:
Created TruncatedIQM class and its firts unit tests.

ywwg committed:
Fix spinny covers not being centered if they aren't square.

ferranlala committed:
Added test double9. Fixed couple of bugs in TruncatedIQM.

ywwg committed:
LateNight: Fix misplaced connection

ferranlala committed:
Added test integers9 to TruncatedIQM. Fixed TruncatedIQM's test doubles9, now checks all 15 values. Changed TruncatedIQM's tests to stream error messages to the EXPECTs rather than qDebug().

ferranlala committed:
Fixed truncatediqm_test.cpp's EXPECTs error messages.

ferranlala committed:
Renamed TruncatedIQM to MovingTruncatedIQM.

ferranlala committed:
Removed unnecessary <QDebug> include in movingtruncatediqm.cpp.

ferranlala committed:
TapFilter now uses MovingTruncatedIQM.

ferranlala committed:
DlgTrackInfo now uses TapFilter to manage tapped bpm.

ferranlala committed:
Fixed TapFilter's lenght harcoded to 4. Fixed MovingTruncatedIQM bug.

ferranlala committed:
Changed DlgTrackInfo bpm tap moving average lenght to 80.

ferranlala committed:
Changed name of MovingTruncatedIQM to MovingInterquartileMean.

ferranlala committed:
Merge branch 'master' into bpm-improvements

kain88-de commented on build/depends.py:
```diff
@@ -902,6 +902,7 @@ def sources(self, build):
                    "encoder/encodervorbis.cpp",
 
                    "tapfilter.cpp",
+                   "movinginterquartilemean.cpp",
```
it looks like you added this file but forgot to add it to git. Please add this file and the corresponding header file. Then we can test your changes.

ferranpujolcamins replied:
Oops, sorry. I've just pushed that.

ferranlala committed:
Added MovingInterquartileMean forgotten files.

ferranlala committed:
Merge branch 'master' into bpm-improvements

rryan commented on src/movinginterquartilemean.h:
```diff
@@ -0,0 +1,44 @@
+#include <QLinkedList>
+#include <QQueue>
+
+//! Truncated Interquartile mean
```
Sorry in advance for nit-pick code style comments. :)

Please remove the exclamation points in front of your comments.

rryan commented on src/movinginterquartilemean.h:
```diff
@@ -0,0 +1,44 @@
+    double m_dMean;
+    int m_iListMaxSize;
+    //! The list keeps input doubles ordered by value.
+    QLinkedList<double> m_List;
```
For members without hungarian please don't capitalize the first letter (e.g. ```m_list``` instead of ```m_List```)

rryan commented on src/dlgtrackinfo.cpp:
```diff
@@ -461,26 +462,12 @@ void DlgTrackInfo::slotBpmThreeFourth() {
-        temp /= kFilterLength;
-        spinBpm->setValue(temp);
+void DlgTrackInfo::slotBpmTap(double averageLength, int numSamples) {
+    double averageBpm = 60.0 * 1000.0 / averageLength;
```
Please guard against division by zero.

rryan commented on src/movinginterquartilemean.cpp:
```diff
@@ -0,0 +1,90 @@
+double MovingInterquartileMean::mean() {
+    if (m_bChanged) {
+        m_bChanged = false;
+        if (m_List.size() <=4) {
```
In general we pad operators with spaces on either side.

```m_list.size() <= 4```

rryan commented on src/movinginterquartilemean.h:
```diff
@@ -0,0 +1,44 @@
+//! and the highest 25% values in order to reduce sensitivity to outliers.
+//!
+//! http://en.wikipedia.org/wiki/Interquartile_mean
+class MovingInterquartileMean {
```
It would be really nice if this (and tapfilter, but I realize the code was already this way) was in src/util. We can move it later though.

ferranlala committed:
Moved movinginterquartilemean.h /.cpp and tapfilter.h / .cpp to src/util.
Guarded division by 0 in DlgTrackInfo::slotBpmTap.
Renamed MovingInterquartileMean::m_List to m_list and m_Queue to m_queue.
Removed ! from movinginterquartilemean.h's comments.
Minor cosmetic code improvements.

ferranlala committed:
Merge branch 'master' into bpm-improvements

ferranpujolcamins commented:
This should address this first issues.

ferranpujolcamins commented:
By the way, what should I do with files I modified that had an author?

kain88-de commented:
> Shall I replace the original author with myself in cases where i completely changed the content of the file?

You can remove the author section completely. There are still some left from the early days. Now we have git to track the author of every line.

ferranlala committed:
Removed author section in some files.

ferranlala committed:
Merge branch 'master' into bpm-improvements

ferranlala committed:
Removed unnecessary include for tapfilter.h in engine/keycontrol.h.

ferranlala committed:
Merge branch 'master' into bpm-improvements

kain88-de commented on src/util/movinginterquartilemean.cpp:
```diff
@@ -0,0 +1,88 @@
+            double interQuartileRange = 2 * quartileSize;
+            QLinkedList<double>::iterator i = m_list.begin() + quartileSize;
+            double d_sum = 0;
+            for (int k = 0; k < 2 * quartileSize; ++k) {
```
you can iterate over two variables at the same time. It would be better here to do that instead of the loop body. This makes the code a bit more verbose.

```for(int k=0, i=m_list.begin()+quartileSize; k<2*quartileSize; ++i, ++k ) {
d_sum += *i;
```

and maybe change `i` to `el` or 'it'. Single letter variables usually indicate that they are used as an index. But iterators aren't an index.

kain88-de commented on src/util/movinginterquartilemean.cpp:
```diff
@@ -0,0 +1,88 @@
+            m_dMean = d_sum / interQuartileRange;
+        } else {
+            // http://en.wikipedia.org/wiki/Interquartile_mean#Dataset_not_divisible_by_four
+            // TODO(Ferran Pujol): Make this more clear.
```
Please remove the TODO before the merge.

kain88-de commented on src/util/movinginterquartilemean.cpp:
```diff
@@ -0,0 +1,88 @@
+            // TODO(Ferran Pujol): Make this more clear.
+            double quartileSize = m_list.size() / 4.0;
+            double interQuartileRange = 2 * quartileSize;
+            int nFullValues = m_list.size() - 2*(int)quartileSize - 2;
```
plese use C++ style casts `static_cast<int>`

kain88-de commented on src/util/movinginterquartilemean.cpp:
```diff
@@ -0,0 +1,88 @@
+            QLinkedList<double>::iterator i = m_list.begin() + (int)quartileSize;
+            double d_sum = *i * quartileWeight;
+            ++i;
+            for (int k = 0; k < nFullValues; ++k) {
```
same comment as above

kain88-de commented on src/util/movinginterquartilemean.cpp:
```diff
@@ -0,0 +1,88 @@
+    m_bChanged = true;
+
+    // Insert new value
+    if (m_list.isEmpty()) {
```
you may want to use a QMap. It already stores the keys in order for you. You just need to insert a garbage value. You have later easy access to the sorted keys in a a QList.

kain88-de commented:
Sorry for the late note. Github doesn't notify us if new commits are added to a PR we only see new comments.

I added a few comments that should help you to make the code more readable.

ferranlala committed:
Make a comment more clear.

ferranlala committed:
Renamed iterators to "it".

ferranlala committed:
Moved it updating into the loop header.

ferranlala committed:
Removed TODO.

ferranlala committed:
Change casts to c++ style.

ferranpujolcamins commented:
Issues should be fixed. Is it necessary to use QMap? The insert code is not that complex.

kain88-de commented on src/test/movinginterquartilemean_test.cpp:
```diff
@@ -0,0 +1,126 @@
+                         0.0573442311111, 0.250574553889, 0.252138432222};
+    for (int i=0; i<15; ++i) {
+        double mean = iqm.insert(input[i]);
+        //TODO(Ferran Pujol): Why does EXPECT_DOUBLE_EQ fail here?
```
This is due to the nature of floating point numbers. They accumulate small round off errors. In this case it should be enough for us if they don't pass with the border you have chosen.

kain88-de commented:
did you test how good the results are with a array size of 80? I can see the BPM drifting a lot getting larger and larger (or smaller and smaller). The drift is somewhat stabilized when I reduce the array size (15).

ferranlala committed:
Removed unwanted comment from a test fixture template.

ferranpujolcamins commented:
I only made the test you've seen. I've calculated the results manually, but I see Matlab has a truncated mean function implemented (trimmean), so I could generate a couple of tests with a massive amount of numbers.

However, if I tap more than 80 beats to a 4/4 song at 125 bpm I get results accurate up to 0.2 bpm. Can you give more details about this drift you saw?

kain88-de commented on src/dlgtrackinfo.cpp:
```diff
@@ -461,26 +459,13 @@ void DlgTrackInfo::slotBpmThreeFourth() {
-        }
-        temp /= kFilterLength;
-        spinBpm->setValue(temp);
+void DlgTrackInfo::slotBpmTap(double averageLength, int numSamples) {
```
numSamples isn't used

kain88-de commented on src/dlgtrackinfo.h:
```diff
@@ -41,7 +40,7 @@ class DlgTrackInfo : public QDialog, public Ui::DlgTrackInfo {
     void trackUpdated();
     void fetchTag();
-
+    
```
you introduced some trailing whitespaces here.

kain88-de commented on src/util/tapfilter.h:
```diff
@@ -18,9 +21,8 @@ class TapFilter : public QObject {
-    int m_iFilterLength;
-    int m_iValidPresses;
+    MovingInterquartileMean m_mean;
+    const int m_iMeanWindowSize;
```
what is the use for this variable? I can't see it used anywhere.

kain88-de commented:
Besides the nitpick comments it looks fine to me.

I couldn't reproduce the  drift today. I guess my own timing was a bit off yesterday.

Thanks for the work

ferranlala committed:
Removed trailing whitespaces.

rryan commented:
Hi @ferranpujolcamins -- looks like this doesn't merge cleanly (it's a minor fix). Could you update it?

rryan commented:
Thanks! Merged.
